snuba/query/expressions.py

```diff
@@ -139,6 +143,21 @@ def accept(self, visitor: ExpressionVisitor[TVisited]) -> TVisited:
         return visitor.visitColumn(self)
 
 
+@dataclass(frozen=True)
+class NestedColumn(Column):
```

Comment:
I went back and forth a dozen of times on how to parse tags[key] into the AST before tags processing. The options were:
- Ignore it, let the parser create a column like `Column(None, "tags[key]", None)` and let the processor find those columns from their name
- Turn them into functions `FunctionCall(None, 'tags', (Literal(None, 'key'),)) and hope for the best in case of naming conflicts since the processor would not know if the user provided `tags[key]` or an actual function call named tags
- Parse the expressions `tags[key]` into a subclass of columns that mke it clear what they were.

I landed on the third by exclusion: I did not like the risk of ambiguity of the other two options and letting the parser just build a simple Column for the tag seemed like lowering the level of abstraction loosing the nuance of what the user wanted during parsing to infer it back during query processing. This seemed off.

Unfortunately this syntax should only be supported for datasets that support tags/contexts (or, more in general, dictionary style nested columns), though we have this as part of the general snuba syntax so, at the point when we do parsing we cannot tell whether the dataset knows about tags.  I think the issue is having this special syntax to begin with instead of using the snuba function call syntax.

Reply:
I think this name is confusing, this is particular to our use of nested columns as representations of mappings for these types which is more restrictive than the general ClickHouse concept of nested columns

Reply:
What about MappingColumn or DictColumn ?